juju/juju
# Improve private OCI registry for caas-image-repo

- Ensure ecr private registry is used before docker registry. Until juju 3.3, we can't remove the default docker registry (i.e. using "jujusolutions" as a repo name instead of the fully qualified name "docker.io/jujusolutions"). This fixes an ordering issue with ecr private (and other repos where you can specify the repo without a subpath) being interpreted as a docker.io repos.
- Once you are using a private OCI registry for caas-image-repo, it is impossible to update the authentication details. This PR allows you to change those authentication details, but not change the registry/repository.

## QA steps

- Create private ECR registry
- Copy jujudb image + push jujud operator image
- Make sure to bootstrap with json repo info like:
- `juju bootstrap k8s --config caas-image-repo='{"repository":"12345.dkr.ecr.eu-west-1.amazonaws.com", "username":"AWS_KEY_ID", "password": "AWS_SECRET_KEY", "region":"eu-west-1" }'`
- Try updating the username and password:
- `juju controller-config caas-image-repo='{"repository":"12345.dkr.ecr.eu-west-1.amazonaws.com", "username":"NEW_AWS_KEY_ID_NEW", "password": "NEW_AWS_SECRET_KEY_NEW", "region":"eu-west-1" }'`
- Check image pull secrets are updated.
- Also try bootstrapping with invalid private ecr `juju bootstrap k8s --config caas-image-repo=12345.dkr.ecr.eu-west-1.amazonaws.com`. It should error `ERROR constructing controller config: empty credential for elastic container registry`.

## Documentation changes

N/A

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2037744

**Jira card:** JUJU-4707
